polkadot/node/network/collator-protocol/src/validator_side/mod.rs

```diff
@@ -1050,6 +1050,47 @@ async fn second_unblocked_collations<Context>(
 	}
 }
 
+fn ensure_seconding_limit_is_respected(
+	relay_parent: &Hash,
+	para_id: ParaId,
+	state: &State,
+) -> std::result::Result<(), AdvertisementError> {
+	let paths = state.implicit_view.paths_via_relay_parent(relay_parent);
+
+	gum::trace!(
+		target: LOG_TARGET,
+		?relay_parent,
+		?para_id,
+		?paths,
+		"Checking seconding limit",
+	);
+
+	for path in paths {
+		let mut cq_state = ClaimQueueState::new();
+		for ancestor in path {
+			let seconded_and_pending = state.seconded_and_pending_for_para(&ancestor, &para_id);
+			cq_state.add_leaf(
+				&ancestor,
+				&state
+					.per_relay_parent
+					.get(&ancestor)
+					.ok_or(AdvertisementError::RelayParentUnknown)?
+					.assignment
+					.current,
+			);
+			for _ in 0..seconded_and_pending {
+				cq_state.claim_at(&ancestor, &para_id);
+			}
+		}
+
+		if !cq_state.can_claim_at(relay_parent, &para_id) {
```

Comment:
Also is it ever possible that on one path the order of paraid claims is different than the other path? i.e.:

```
0       1    2
A ----> B -> B
   \
     -> A -> A
```

If that would ever happen it can cause a deadlock where we ignore all advertisements. All adverts for para B would be rejected because they cannot be claimed in bottom fork while all A adverts rejected because they don't fit into upper fork.

As far as I know this should never happen but just want to make absolutely sure.

Reply:
As agreed - we will accept the candidate if there is a free claim queue slot for it at any of the forks.